server/src/main/java/org/apache/kafka/server/transaction/AddPartitionsToTxnManager.java

```diff
@@ -243,11 +243,10 @@ public void addOrVerifyTransaction(
                     Collectors.toMap(Function.identity(), tp -> Errors.COORDINATOR_NOT_AVAILABLE)));
         } else {
             AddPartitionsToTxnTopicCollection topicCollection = new AddPartitionsToTxnTopicCollection();
-            topicPartitions.stream().collect(Collectors.groupingBy(TopicPartition::topic)).forEach((topic, tps) -> {
-                topicCollection.add(new AddPartitionsToTxnTopic()
-                        .setName(topic)
-                        .setPartitions(tps.stream().map(TopicPartition::partition).collect(Collectors.toList())));
-            });
+            topicPartitions.stream().collect(Collectors.groupingBy(TopicPartition::topic)).forEach((topic, tps) ->
```

Comment:
Please consider leveraging the topic name index to avoid regrouping. for example:
```java
            topicPartitions.forEach(tp -> {
                var r = topicCollection.find(tp.topic());
                if (r == null) {
                    r = new AddPartitionsToTxnTopic().setName(tp.topic());
                    topicCollection.add(r);
                }
                r.partitions().add(tp.partition());
            });
```

Reply:
Updated, thanks!